### ISSUE TITLE
S3: Discuss ways to list and tally objects efficiently

### ISSUE THREAD

**Issue by jlevy** [open-guides/og-aws#58]
Topics:
- Need for multi-threaded crawl
- Prefix-based listings, with separators
- Maybe: Using markers to optimize if alphabet is not known

https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGET.html

**Comment by adrianbg** [comment]
From Slack chat:

adrian [10:21 AM]  
start a bunch of workers, any time one of them is idle, pick one of the workers that isn’t idle and split its work between them. so if non-idle worker is iterating over prefix /a/b/c, then assign /a/b/c/<middle of alphabet> to the idle worker, and tell the non-idle worker to stop when it reaches that point. it may be that the new task you assign to the idle worker is trivial but then it just becomes idle again

[10:22]  
so there’s some blind exploration of the tree but it’s always in a binary search mode

[10:22]  
and you can know the alphabet.. the alphabet is what.. bytes? ascii?

adrian [10:40 AM]  
http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGET.html — the `marker` parameter means you can binary search through prefixes without separately testing each character. if you do a list for /prefix/<middle> and it’s empty then you know that [<middle>, <end>] is empty (edited)

[10:42]  
you could do a search over all byte strings but then only actually follow paths that’re valid utf-8. then you’d be able to take advantage of utf-8’s optimization for latin strings while still supporting chinese keys.

jlevy [10:44 AM]  
Haha, I guess there are a few hacks to do. But I don’t think you can do byte prefixes, only Unicode char prefixes.

adrian [10:44 AM]  
yeah

[10:45]  
my solution is handwavey and I hope for my sake that it stays that way

jlevy [10:45 AM]  
So it’d still be binary search over all Unicode, which I’m guesstimating adds a lot of overhead. Better to just require “mostly ascii keys" perhaps.

adrian [10:46 AM]  
so the first branch point in a given search is “is the next character <= 127"

[10:48]  
it’s not “is the next character <= # unicode characters / 2"

[10:49]  
think of it like a binary search over strings of _bits_ in which you only actually test a string if it happens to be valid utf-8

**Comment by jlevy** [comment]
@ThanosBaskous or @weirded perhaps you have further thoughts?

**Comment by ThanosBaskous** [comment]
I haven't had to do this myself but maybe something like the first half of [this](https://blogs.aws.amazon.com/bigdata/post/Tx1XU0OQAZER3MI/Turning-Amazon-EMR-into-a-Massive-Amazon-S3-Processing-Engine-with-Campanile) would work if you have a well-distributed keyspace (use EMR / hadoop to extract the full list)? Or is this heavier weight than you are looking for?

I think you touch on something similar in the last bullet in the original issue description.

**Comment by kimschmidtsbrain** [comment]
Found  open source code that takes your S3 bucket items & make them into a website for downloading & archival for non-tech people. Would that be of interest? I locked down access via S3 policy limiting access to certain IPs

**Comment by weirded** [comment]
The binary search approach is the best we could think of, as well - but obviously, it requires knowing your character set and general make up of the keys in the bucket, so building it as a generic library is hard. 

Another approach is to use a secondary store (such as DynamoDB or RDS) to keep track of the keys. You can use S3 Event notifications to keep the store updated, possibly with a "slow background scan" to re-sync periodically. Those stores are better optimized for enumerating lots of entries than S3.

**Comment by kaihendry** [comment]
I successfully used https://www.npmjs.com/package/s3ls in a lambda function to list items in my S3 bucket.

**Comment by adrianbg** [comment]
@weirded the character set is UTF-8. It should be possible to recursively subdivide the set of valid UTF-8 strings using the ListBucket operations' `marker` parameter so you can do the search efficiently without any extra knowledge of the key distribution in the bucket.

**Comment by weirded** [comment]
@adrianbg yeah though most naming schemes I've seen use ASCII only, which can help narrow things down a lot more quickly than UTF-8. In our internal use case, the first 16 characters were hex strings, so we had only 16 characters to go by :)

**Comment by adrianbg** [comment]
@weirded Yeah, mine were hex too. Probably like you I just broke up the listing by the first hexit. Didn't think of this binary search idea until later.

Actually, binary search wouldn't be bad at all, even with hex keys. Here's a quick example where the numbers are binary prefixes. The first few queries I would do are:

- LIST /00000000 (starts listing ASCII digits -- remember that S3 automatically starts listing at the first key after the `marker` parameter)
- LIST /10000000 (ie. are there any keys with 1 as the first bit? This would return no results, ruling out Unicode thanks to UTF-8)
- LIST /01000000 (starts listing ASCII letters)

Then you'd have two jobs each listing a large set of keys. If you want to use more threads then you could keep subdividing the work.
